File: src/main/java/com/itszuvalex/femtocraft/managers/assistant/ManagerAssistant.java
package com.itszuvalex.femtocraft.managers.assistant;

import com.itszuvalex.femtocraft.Femtocraft;
import com.itszuvalex.femtocraft.utils.FemtocraftFileUtils;
import net.minecraft.nbt.CompressedStreamTools;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.world.World;
import org.apache.logging.log4j.Level;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FilenameFilter;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

public class ManagerAssistant {
    private static final String dir = "Assistants";
    private static final String assistKey = "assistants";
    private static HashMap<String, Map<String, AssistantPermissions>> data;

    public ManagerAssistant() {
        data = new HashMap<String, Map<String, AssistantPermissions>>();
    }

    public Map<String, AssistantPermissions> getPlayerAssistants(String owner) {
        return data.get(owner);
    }

    public boolean isPlayerAssistant(String owner, String user) {
        Map<String, AssistantPermissions> pdata = data.get(owner);
        return (pdata != null) && (pdata.keySet().contains(user));
    }

    public boolean addAssistantTo(String owner, String user) {
        Map<String, AssistantPermissions> pdata = data.get(owner);
        if (pdata == null) {
            pdata = addPlayerAssistant(owner);
        }
        if (pdata != null) {
            return pdata.put(user, new AssistantPermissions(user)) == null;
        }
        return false;
    }

    public Map<String, AssistantPermissions> addPlayerAssistant(String uname) {
        if (data.containsKey(uname)) {
            return data.get(uname);
        } else {
            Map<String, AssistantPermissions> pdata = new TreeMap<String, AssistantPermissions>();
            data.put(uname, pdata);
            return pdata;
        }
    }

    public boolean removeAssistantFrom(String owner, String user) {
        Map<String, AssistantPermissions> pdata = data.get(owner);
        if (pdata != null) {
            return pdata.remove(user) != null;
        }
        return false;
    }

    public boolean save(World world) {
        try {
            File file = new File(FemtocraftFileUtils.savePathFemtocraft(world), dir);
            if (!file.exists()) {
                file.mkdirs();
            }
            for (String username : data.keySet()) {
                try {
                    File userfile = new File(file, username + ".dat");
                    if (!userfile.exists()) {
                        userfile.createNewFile();
                    }

                    NBTTagCompound gTag = new NBTTagCompound();
                    NBTTagList pDataTag = new NBTTagList();

                    NBTTagList assistTag = new NBTTagList();
                    for (String assistant : data.get(username).keySet()) {
                        NBTTagCompound assistantTag = new NBTTagCompound();
                        data.get(username).get(assistant).saveToNBT(assistantTag);
                        assistTag.appendTag(assistantTag);
                    }

                    gTag.setTag(assistKey, assistTag);

                    FileOutputStream fos = new FileOutputStream(userfile);
                    CompressedStreamTools.writeCompressed(gTag, fos);
                    fos.close();
                } catch (Exception e) {
                    Femtocraft.log(Level.ERROR,
                            "Failed to save assistant data in world - "
                            + FemtocraftFileUtils.savePathFemtocraft(world) + " for player" + username + "."
                    );
                    e.printStackTrace();
                }
            }

        } catch (Exception e) {
            Femtocraft.log(Level.ERROR,
                    "Failed to save assistant data in world - "
                    + FemtocraftFileUtils.savePathFemtocraft(world) + "."
            );
            e.printStackTrace();
            return false;
        }
        return true;
    }

    public boolean load(World world) {
        try {
            File file = new File(FemtocraftFileUtils.savePathFemtocraft(world), dir);
            if (!file.exists()) {
                Femtocraft.log(Level.WARN, "No assistant data"
                                           + " found for world - " + FemtocraftFileUtils.savePathFemtocraft(world) +
                                           ".");
                return false;
            }

            data.clear();

            for (File pfile : file.listFiles(new FilenameFilter() {
                @Override
                public boolean accept(File dir, String name) {
                    return name.endsWith(".dat");
                }
            }))

            {
                String username = pfile.getName().substring(0, pfile.getName().length() - 4);
                try {
                    FileInputStream fis = new FileInputStream(pfile);
                    NBTTagCompound gTag = CompressedStreamTools.readCompressed(fis);
                    fis.close();
                    NBTTagList assistTag = gTag.getTagList(assistKey, 10);

                    for (int j = 0; j < assistTag.tagCount(); j++) {
                        NBTTagCompound permTag = assistTag.getCompoundTagAt(j);
                        AssistantPermissions perm = new AssistantPermissions();
                        perm.loadFromNBT(permTag);
                        Map<String, AssistantPermissions> pdata = data.get(username);
                        if (pdata == null) {
                            pdata = addPlayerAssistant(username);
                        }
                        if (pdata != null) {
                            pdata.put(perm.assistant, perm);
                        }
                    }
                } catch (Exception e) {
                    Femtocraft.log(Level.ERROR,
                            "Failed to load assistant data in world - "
                            + FemtocraftFileUtils.savePathFemtocraft(world) + " for player" + username + "."
                    );
                    e.printStackTrace();
                }
            }

        } catch (
                Exception e
                )

        {
            Femtocraft.log(Level.ERROR,
                    "Failed to load assistant data in world - "
                    + FemtocraftFileUtils.savePathFemtocraft(world) + "."
            );
            e.printStackTrace();
            return false;
        }

        return true;
    }
}

File: src/main/scala/com/itszuvalex/femtocraft/managers/assistant/ManagerAssistant.scala
package com.itszuvalex.femtocraft.managers.assistant

import java.io.{FilenameFilter, File, FileInputStream, FileOutputStream}
import java.util

import com.itszuvalex.femtocraft.Femtocraft
import com.itszuvalex.femtocraft.api.core.Configurable
import com.itszuvalex.femtocraft.api.managers.IAssistantManager
import com.itszuvalex.femtocraft.api.utils.FemtocraftFileUtils
import net.minecraft.nbt.{CompressedStreamTools, NBTTagCompound, NBTTagList}
import net.minecraft.world.World
import org.apache.logging.log4j.Level

import scala.collection.JavaConversions._
import scala.collection.mutable

@Configurable object ManagerAssistant extends IAssistantManager {
  private               val dir              = "Assistants"
  private               val assistKey        = "assistants"
  private               val data             = new mutable.HashMap[String, mutable.HashMap[String, AssistantPermissions]]
  private               var lastWorld: World = null
  @Configurable private val debugMessages    = false


  override def isPlayerAssistant(owner: String, user: String) = getPlayerAssistants(owner).contains(user)

  override def addAssistantTo(owner: String, user: String): Unit = {
    getPlayerAssistantsPermissions(owner).put(user, new AssistantPermissions(user))
    save()
  }

  override def removeAssistantFrom(owner: String, user: String): Unit = {
    getPlayerAssistants(owner).remove(user)
    save()
  }

  override def getPlayerAssistants(owner: String): util.Collection[String] =
    data.getOrElseUpdate(owner, new mutable.HashMap[String, AssistantPermissions]).keySet

  def save(): Boolean = if (lastWorld == null) false else save(lastWorld)

  def save(world: World): Boolean = {
    try {
      val file = new File(FemtocraftFileUtils.savePathFemtocraft(world), dir)
      if (!file.exists) {
        file.mkdirs
      }
      data.keySet.view.foreach { username =>
        try {
          val userfile = new File(file, username + ".dat")
          if (!userfile.exists) {
            userfile.createNewFile
          }
          val gTag = new NBTTagCompound
          val assistTag = new NBTTagList
          getPlayerAssistantsPermissions(username).values.foreach { perms =>
            val assistantTag = new NBTTagCompound
            perms.saveToNBT(assistantTag)
            assistTag.appendTag(assistantTag)
                                                                  }
          gTag.setTag(assistKey, assistTag)
          val fos = new FileOutputStream(userfile)
          CompressedStreamTools.writeCompressed(gTag, fos)
          fos.close()
        }
        catch {
          case e: Exception =>
            Femtocraft
            .log(Level.ERROR,
                 "Failed to save assistant data in world - " + FemtocraftFileUtils
                                                               .savePathFemtocraft(world) + " for player" + username + ".")
            e.printStackTrace()
        }
                               }
    }
    catch {
      case e: Exception =>
        Femtocraft
        .log(Level.ERROR,
             "Failed to save assistant data in world - " + FemtocraftFileUtils.savePathFemtocraft(world) + ".")
        e.printStackTrace()
        return false
    }
    if (debugMessages) Femtocraft.log(Level.INFO, "Saving Assistant data for world - " + FemtocraftFileUtils.savePathFemtocraft(world) + ".")
    true
  }

  def load(world: World): Boolean = {
    save()
    lastWorld = world
    try {
      val file = new File(FemtocraftFileUtils.savePathFemtocraft(world), dir)
      if (!file.exists) {
        Femtocraft
        .log(Level.WARN,
             "No assistant data" + " found for world - " + FemtocraftFileUtils.savePathFemtocraft(world) + ".")
        return false
      }
      else {
        if (debugMessages) Femtocraft.log(Level.INFO, "Loading Assistant data for world - " + FemtocraftFileUtils.savePathFemtocraft(world) + ".")
      }
      data.clear()
      for (pfile <- file.listFiles(new FilenameFilter() {
        def accept(dir: File, name: String): Boolean = {
          name.endsWith(".dat")
        }
      }).view) {
        val username = pfile.getName.substring(0, pfile.getName.length - 4)
        try {
          val fis = new FileInputStream(pfile)
          val gTag = CompressedStreamTools.readCompressed(fis)
          fis.close()
          val assistTag = gTag.getTagList(assistKey, 10)

          for (j <- 0 until assistTag.tagCount) {
            val permTag = assistTag.getCompoundTagAt(j)
            val perm = new AssistantPermissions
            perm.loadFromNBT(permTag)
            getPlayerAssistantsPermissions(username).put(perm.assistant, perm)
          }
        }
        catch {
          case e: Exception =>
            Femtocraft
            .log(Level.ERROR,
                 "Failed to load assistant data in world - " + FemtocraftFileUtils
                                                               .savePathFemtocraft(world) + " for player" + username + ".")
            e.printStackTrace()
        }
      }
    }
    catch {
      case e: Exception =>
        Femtocraft
        .log(Level.ERROR,
             "Failed to load assistant data in world - " + FemtocraftFileUtils.savePathFemtocraft(world) + ".")
        e.printStackTrace()
        return false
    }
    true
  }

  def getPlayerAssistantsPermissions(owner: String) = data.getOrElseUpdate(owner, new
      mutable.HashMap[String, AssistantPermissions])
}
